### DISCUSSION TITLE
Welcome to unify-flowret Discussions!

### DISCUSSION THREAD

**Discussion by deepakarora3** [americanexpress/unify-flowret#8] [General]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by dunglt-iks** [comment]
Hello Deepak Arora,
I am BA guys, but I am interesting in coding :). 
I am very interest with Flowret workflow implementation. 
I have question below. 

Assumption that we have flow
A -> B (p_route)
B -> Branch1 (pause)
B -> Branch2 (pause)
Branch1-> C (join)
Branch2 -> C (join)
C -> End.

When flow is executed. It will be stopped and wait for resume.
Currently flow will be pended at branch1 or branch2 (it depends on which thread completes execution first).
We assume that flow is pended at branch1. But, we would like to resume from branch2.
**How does current implementation support above scenario?**
In reality, Branch1 represents for Manager1 approval. Branch2 represents for Manager2 approval in parallel, no matter which approval finishes first. Then flow will go to next step if and only if both approvals are done.

**Reply by deepakarora3** [reply]
Hi - thanks for trying things out. A couple of clarifications. You should only be pending on steps. So assuming that Branch1 and Branch2 are steps. To your question, if we do parallel processing and if multiple paths pend at the same time, then the application will pend at the step which pended first. Then when the application is resumed, it will start from that step and after that step reaches join, it will check if there are any other pends. If there are then the application will pend at that step. Once it is resumed from there and goes to join, then the application will move ahead. So in your case, if the pend happens for both managers, the first manager approves, then the second manage approves and then the flow goes beyond the join. This is explained in the documentation too. Reproducing the content below:

**What happens if multiple branches in a parallel process pend?**
In the case of the parallel flow described previously, it is possible for more than one branch to pend. Flowret, in such a case will record all pends but will return the pend instance that occurred first back to the application as part of the ON_PROCESS_PEND event.
When the condition for this pend has cleared and the process is resumed, Flowret will first check if there are any other pends that have not yet been returned to the application. If there are, the next pend will be returned immediately without further execution on any branch till the last pend is resolved.
In future versions, a feature may be provided for Flowret to return a list of pends to the application.

Hope this helps.

**Reply by dunglt-iks** [reply]
Dear Deepak Arora,

Thank you for reply, I understand your explanation. I also tried it out. But, my concern is another thing. Branch1 and Branch2 are paused and wait for corresponding Manager approval. Let assume flow is pended at Branch1 (because it were ran before Branch2). But Manager2 initiate his/her approval before Manager1. So, Flowret will resume process from Branch1 and subsequence steps on Branch1's path will be executed. It is not our expectation. We would like to resume from Branch2, and its subsequence steps should be executed. 

I think we should add feature to resume from specific pended path, and property execPath in ProcessInfo should be a list. It can contain more than one pendPath. When we call resumeCase method, I should pass which path should be resumed from.

What do you think?
Thank you!

**Reply by deepakarora3** [reply]
I see what you mean - unfortunately we do not have that feature - as of now we can be pended only at one step at any time - though we can certainly think about it. There are quite a few other things that will need to be taken care of. Let me think about this for some time.

**Reply by dunglt-iks** [reply]
Hi Deepak Arora,
Thank you for considering. I am trying modify code for resolving the issue :).
What are other things that you are taking care of? I am considering to use the framework for my project. And I would like to know if there is any difficulty/un-support feature from the framework. Or I can contribute my idea/effort/fixing bug :)

Hope getting your reply!!!

**Comment by mingardia** [comment]
Hi Deepak, I am looking for a solution and like what I see in this project.  The main hesitation I have is support for versioned workflows which is a common issue.  So if have a large number of workflows running and we need to change one we want the existing executions to continue on the journey they are on, but new executions would take up the new definition.  Is this possible, or something that needs to be added to the system?

**Reply by deepakarora3** [reply]
Hi Michael - thanks for your interest and I am glad you like what you see. As regards your question on supporting versioned workflows, the feature already exists. When Flowret is asked to start a case, it is provided a process definition. Flowret locks this process definition to the case for the lifetime of the case. If tomorrow you were to change the process definition, the new cases would start to use the new definition whereas the inflight cases would continue to use the old definition. Specifically, it writes a file by the name of flowret_journey-<case_id> to store this information. I have mentioned this in the documentation as well - in the sections "Capability and feature summary" and in "Start a case". Happy to answer more queries. Thanks.

**Reply by mingardia** [reply]
Perfect !!!.  Sounds like exactly what we are looking for.  I did a clone and build it via mvn package .. there are multiple unit tests that are failing, is that normal for the main branch:  [ERROR] Tests run: 26, Failures: 20, Errors: 0, Skipped: 0

**Reply by deepakarora3** [reply]
Really? I just ran TestFlowret in test_singlular package and everything looks fine to me. I am running on Windows. Are you on a Mac? Can you just run the test case testClean() and let me know what error you are seeing? It may have to do with how the text files are stored on Mac - to be honest I have tested these only on Windows. Let me also have someone test it here on Mac and report back. Rest assured all are working and the software is as updated as can be - we are using it internally all the time :)

**Comment by mingardia** [comment]
Yea on a M2 mac, I forked it and will debug and create a pull request for what I find.


Exception -> Cannot resume a case that has already completed. Case id -> 1
>
        at org.junit.jupiter.api.AssertionUtils.fail(AssertionUtils.java:54)
        at org.junit.jupiter.api.AssertEquals.failNotEqual(AssertEquals.java:195)
        at org.junit.jupiter.api.AssertEquals.assertEquals(AssertEquals.java:184)
        at org.junit.jupiter.api.AssertEquals.assertEquals(AssertEquals.java:179)
        at org.junit.jupiter.api.Assertions.assertEquals(Assertions.java:508)
        at com.americanexpress.unify.flowret.test_singular.TestFlowret.myAssertEquals(TestFlowret.java:236)
        at com.americanexpress.unify.flowret.test_singular.TestFlowret.testResume(TestFlowret.java:361)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:568)
        at org.junit.platform.commons.util.ReflectionUtils.invokeMethod(ReflectionUtils.java:515)
        at org.junit.

Setting up the project now to debug

Mike

From: Deepak Arora ***@***.***>
Reply-To: americanexpress/unify-flowret ***@***.***>
Date: Wednesday, February 7, 2024 at 12:22 PM
To: americanexpress/unify-flowret ***@***.***>
Cc: Michael Ingardia ***@***.***>, Comment ***@***.***>
Subject: Re: [americanexpress/unify-flowret] Welcome to unify-flowret Discussions! (Discussion #8)


Really? I just ran TestFlowret in test_singlular package and everything looks fine to me. I am running on Windows. Are you on a Mac? Can you just run the test case testClean() and let me know what error you are seeing? It may have to do with how the text files are stored on Mac - to be honest I have tested these only on Windows. Let me also have someone test it here on Mac and report back.

**Comment by mingardia** [comment]
Exception -> Cannot resume a case that has already completed. Case id -> 1

> but was: <Received event -> ON_PROCESS_START, isPendAtSameStep -> false

Received enqueue request. Json below

{

"milestones" : [ {

"name" : "case_start_1",

"setup_on" : "case_start",

"type" : "case_level",

"applied_at_age" : "20d",

"clock_starts" : "immediately",

"action" : "CANCEL:CW"

}, {

"name" : "case_start_2",

"setup_on" : "case_start",

"type" : "case_level",

"applied_at_age" : "30d",

"clock_starts" : "immediately",

"action" : "CANCEL:CW"

} ]

}

What is interesting is does not happen when I run through the IDE but does happen when I run from the command line using mvn package, but running the test in intelij the test passes.  Could be a race condition or that when maven is running its running the tests concurrently perhaps.  Will debug further and get back to you.

Mike

From: Deepak Arora ***@***.***>
Reply-To: americanexpress/unify-flowret ***@***.***>
Date: Wednesday, February 7, 2024 at 12:22 PM
To: americanexpress/unify-flowret ***@***.***>
Cc: Michael Ingardia ***@***.***>, Comment ***@***.***>
Subject: Re: [americanexpress/unify-flowret] Welcome to unify-flowret Discussions! (Discussion #8)


Really? I just ran TestFlowret in test_singlular package and everything looks fine to me. I am running on Windows. Are you on a Mac? Can you just run the test case testClean() and let me know what error you are seeing? It may have to do with how the text files are stored on Mac - to be honest I have tested these only on Windows. Let me also have someone test it here on Mac and report back.

**Comment by deepakarora3** [comment]
Yes – that is it – I have always run unit tests from within IntelliJ. I too will take a look when I get some time. Thanks!

**Comment by mingardia** [comment]
Not sure but it seems have to do something with the way your determining success or failure centric to the output to the console which I would probably try to come up with some other way of doing that as there are variability that can easily be introduced centric to new logs, or other things that would cause the unit test to fail, even though logically really nothing is wrong.  Trying to determine what the difference is right now will let you know if I find something.  In the meantime I will start bringing this into the software I am working on.

http://www.end2endlogic.com
http://www.b2bintegrator.com

Are the two things I am working on

Mike

From: Deepak Arora ***@***.***>
Reply-To: americanexpress/unify-flowret ***@***.***>
Date: Wednesday, February 7, 2024 at 1:55 PM
To: americanexpress/unify-flowret ***@***.***>
Cc: Michael Ingardia ***@***.***>, Comment ***@***.***>
Subject: Re: [americanexpress/unify-flowret] Welcome to unify-flowret Discussions! (Discussion #8)

Yes – that is it – I have always run unit tests from within IntelliJ. I too will take a look when I get some time. Thanks!

Regards,

Deepak Arora

**Comment by deepakarora3** [comment]
Noted. As you can imagine, it is not straight forward to write unit test cases for an orchestrator. The strategy that I had adopted was that I would compare the output of the logs to determine if a test case passed or not. The other way I could have done was to examine the contents of the files created. For some reason, I went with the former. Maybe if I get some time to work on these test cases, I may change the design but as of now it is doing the job albeit via IntelliJ :)

Went through your project links and they indeed seem interesting. It would be great for unify-flowret to participate. I wish you all the very best!

**Comment by mingardia** [comment]
Most likely it’s the difference in carriage return line feed values in the file you’re comparing against vs. what is being generated on the mac.  I suspect it would fail on linux in the same way as a result.  Will play around with it but I think that’s the difference that in the command line your getting /r vs. /n values at the end of the lines.

Mike




From: Michael Ingardia ***@***.***>
Date: Wednesday, February 7, 2024 at 2:11 PM
To: americanexpress/unify-flowret ***@***.***>, americanexpress/unify-flowret ***@***.***>
Cc: Comment ***@***.***>
Subject: Re: [americanexpress/unify-flowret] Welcome to unify-flowret Discussions! (Discussion #8)

Not sure but it seems have to do something with the way your determining success or failure centric to the output to the console which I would probably try to come up with some other way of doing that as there are variability that can easily be introduced centric to new logs, or other things that would cause the unit test to fail, even though logically really nothing is wrong.  Trying to determine what the difference is right now will let you know if I find something.  In the meantime I will start bringing this into the software I am working on.

http://www.end2endlogic.com
http://www.b2bintegrator.com

Are the two things I am working on

Mike

From: Deepak Arora ***@***.***>
Reply-To: americanexpress/unify-flowret ***@***.***>
Date: Wednesday, February 7, 2024 at 1:55 PM
To: americanexpress/unify-flowret ***@***.***>
Cc: Michael Ingardia ***@***.***>, Comment ***@***.***>
Subject: Re: [americanexpress/unify-flowret] Welcome to unify-flowret Discussions! (Discussion #8)

Yes – that is it – I have always run unit tests from within IntelliJ. I too will take a look when I get some time. Thanks!

Regards,

Deepak Arora

**Comment by deepakarora3** [comment]
Hi - just wanted to let you know that 1.6.0 release has a new section - getting started in 5 minutes - in the readme.md documentation. I hope it helps new people wanting to give Flowret a try.

**Comment by deepakarora3** [comment]
Hi all - have released 1.7.1 - no functional changes - a revamp of unit test cases - you can read more about it in the releases section. Thanks

**Reply by markjhardy** [reply]
Is this Engine still in use at American Express and would you consider the product complete or are there future enhancements coming?

I would like to see step parameters as an option so step specific name/value pairs can be easily stored in each step if needed. Was the intent of the UserData in the step meant to pass step specific data into each step?

**Reply by markjhardy** [reply]
At American Express do you persist to a Data Store in your DAO or just to the file system? If a data store do you have a SampleDao.java instance that shows what is expected to read/write to a SQL or NON-SQL data store? Thanks.

**Reply by deepakarora3** [reply]
Hi Mark,

Thanks for reaching out and your interest. The engine is very much being used in American Express. It runs one of our big complex BPM / workflow platforms and has been a great success. It is at present maintained actively by me and we are involving other folks also in this. As you would have seen, the enhancements / changes have thinned down over the past months - this is only due to the fact that it right now has pretty much all that we need and it is stable and battle tested. So from that point of view, it is more or less complete - but we are open to new features and enhancements.

Regarding user data, yes, you are right - that was exactly the intent - there were scenarios that we wanted to use the same step implementation at multiple places in a workflow but wanted it to do something different based on where it was placed in the workflow.

We persist data to a NoSQL database using an implementation of the FlowretDao interface. It is extremely straightforward to use as the case id is expected to be unique and functions as a primary key. It is equally straightforward to write to an RDBMS table using another implementation. I cannot share our implementation but these are the steps - have an application object A that manages connections to a NoSQL. Then have an implementation of FlowretDao interface that has access to object A and uses it to get a connection and then do the read and write. Maybe some time in the future, I will include a dummy implementation.

I would not advise writing to the file system for high volumes unless you can ensure that you distribute your files across multiple folders and each folder does not contain more than a few thousand files - filesystems tend to start going unresponsive once you start putting a lot of files in one folder. What you choose as the data store is really up to you - if you ask me - I would suggest a NoSQL or an RDBMS.

To know what gets written out or read, you can use FileDao.java or SampleFlowretDao.java - both are in the test folder.

Hope this helps.

**Comment by markjhardy** [comment]
Thank you Deepak,That is exactly what I was looking for. I think this will be a great fit for our project. MarkOn Nov 20, 2024, at 7:21 PM, Deepak Arora ***@***.***> wrote:﻿
Hi Mark,
Thanks for reaching out and your interest. The engine is very much being used in American Express. It runs one of our big complicated BPM / workflow platforms and has been a great success. It is at present maintained actively by me and we are involving other folks also in this. As you would have seen, the enhancements / changes have thinned down over the past months - this is only due to the fact that it right now has pretty much all that we need and it is stable and battle tested. So from that point of view, it is more or less complete - but we are open to new features and enhancements.
Regarding user data, yes, you are right - that was exactly the intent - there were scenarios that we wanted to use the same step implementation at multiple places in a workflow but wanted it to do something different based on where it was placed in the workflow.
We persist data to a NoSQL database using an implementation of the FlowretDao interface. It is extremely straightforward to use as the case id is expected to be unique functions as a primary key. It is equally straightforward to write to an RDBMS table using another implementation. I cannot share our implementation but these are the steps - have an application object A that manages connections to a NoSQL. Then have an implementation of FlowretDao interface that has access to object A and uses it to get a connection and then do the read and write. Maybe some time in the future, I will include a dummy implementation.
I would not advise writing to the file system for high volumes unless you can ensure that you distribute your files across multiple folders and each folder does not contain more than a few thousand files - filesystems tend to start going unresponsive once you start putting a lot of files in one folder. What you choose as the data store is really up to you - if you ask me - I would suggest a NoSQL or an RDBMS.
To know what gets written out or read, you can use FileDao.java or SampleFlowretDao.java - both are in the test folder.
Hope this helps.

**Comment by khushaalsharma** [comment]
Hi, I am new to this library for workflow management. I tried to understand how the parallel execution works, though I understand the theory, I am not able to understand how do I implement the join after each path is completed in the parallel flow? Can anyone help with implementation part?

**Reply by deepakarora3** [reply]
Thanks for your interest. In order to see how paralle processing is implemented using parallel route and join constructs, please refer to the file TestFlowretParallel.java in the folder test_parallel in tests. If you want to look up how to define a journey json for parallel processing, you can look up the file parallel_test.json in the test resources folder. Hope this helps. Thanks.

**Comment by Isira-Uditha** [comment]
Hi @deepakarora3 and team,

I've encountered what appears to be a bug in Flowret v1.7.5 when using circular route patterns. I have a ticket lifecycle workflow where a route handler needs to loop back to itself and sometimes wait for user input. The issue happens when a singular route returns `OK_PROCEED` on first execution, then loops back and returns `OK_PEND_EOR` on the second execution.

My workflow flow looks like this:
```
start → update_ticket_status (route) → send_resolved_email (step) → update_ticket_status (loops back)
```

The route handler checks for a process variable `ticketStatus`. If the status exists, it clears the variable and returns `OK_PROCEED` with the appropriate branch (like "RESOLVED"). The workflow then continues to an email step and loops back to the same route. When it loops back and the status is empty, the route returns `OK_PEND_EOR` to wait for the next user input.

Here's exactly what happens:
1. ✅ First call: `ticketStatus="RESOLVED"` → route returns `OK_PROCEED` with branch "RESOLVED"
2. ✅ Email step executes successfully  
3. ✅ Loops back to route, `ticketStatus=""` (cleared)
4. ✅ Route handler returns `OK_PEND_EOR` successfully
5. ❌ Framework throws `UnifyException` at `ExecThreadTask.processSingularRoute(ExecThreadTask.java:651)`

The problem is that while the first cycle works perfectly fine, on the second execution when the route returns `OK_PEND_EOR`, Flowret throws an `UnifyException`. What's strange is that my route handler executes successfully and returns the response without any exceptions - the failure happens inside Flowret's internal processing.

Looking at the source code in `Utils.java`, I noticed that Flowret has special handling for routes where it converts `OK_PROCEED` to `OK_PEND_EOR` internally for re-evaluation. I'm wondering if there's a conflict when a route directly returns `OK_PEND_EOR` in a circular pattern, causing some internal state management issue.

The documentation confirms that routes can use `OK_PEND_EOR`, and my handler creates the response successfully, but something in the framework's processing of this response in circular workflows seems to fail. I've worked around this by converting the circular route to a step handler that uses Flowret's ticket mechanism for navigation, and steps seem to handle `OK_PEND_EOR` and self-loops much better than routes.

Is this a known limitation of circular routes, or should this pattern be supported? I'm using Flowret 1.7.5 with Java 17 and Spring Boot 3.x. Would appreciate any guidance on whether this is expected behavior or if there's a recommended pattern for "wait for input" scenarios in circular workflows.

Thanks for the great work on Flowret! It's been very helpful for our complex workflow requirements, and I hope this feedback helps improve the framework.

**Reply by deepakarora3** [reply]
Hi @Isira-Uditha ,

Thanks for reaching out and providing a detailed analysis. I will look into this at the earliest and let you know. Will try to resolve end of this week or early next. Also, glad to hear that it is is helping you folks!

Regards
Deepak

**Reply by deepakarora3** [reply]
OK - I looked at your scenario - there is one thing which I should I have clarified in the documentation due to which the confusion has occurred. The issue you are facing is because a route is not expected to return either OK_PEND or OK_PEND_EOR. In other words, a route should not be deliberately pending as that is not in its nature. Its nature is to take a decision based on the information available and move execution to one r more of the emanating branches. This has been a deliberate decision for Flowret and I should have clarified in the documentation which I will do in the next release.

In my opinion, you should have a step which is used for pending if the ticket status is cleared. Once the ticket status moves to resolved and you resume the process, it can move to a route which examines what to do next. Something like below:

![1](https://github.com/user-attachments/assets/bb477657-7f5b-4870-8942-e5187e9d1fa4)

Hope this helps. Let me know if you have any questions.
